### PR TITLE
ad7768-1fmcz: Added essential functions to the noos ad77681 library

### PR DESCRIPTION
**Pull request by jgarlitos**

### CODE REVIEW

File: ad7768-1fmcz/ad77681.h [hunk 1]
```diff
@@ -522,28 +525,52 @@ struct ad77681_dev {
...
-	enum ad77681_crc_sel 		crc_sel;
-	uint8_t					status_bit;
+	enum ad77681_power_mode			power_mode;
+	enum ad77681_mclk_div			mclk_div;
+	enum ad77681_conv_mode 			conv_mode;
+	enum ad77681_conv_diag_mux 		diag_mux_sel;
+	bool 							conv_diag_sel;
+	enum ad77681_conv_len			conv_len;
+	enum ad77681_crc_sel 			crc_sel;
+	uint8_t							status_bit;
```
Comment (cristiantpop) [inline ad7768-1fmcz/ad77681.h:535]: Extra tab has been added between type and variable on lines: 525 - 535.
They should probably remain untouched.

File: ad7768-1fmcz/ad77681.h [hunk 2]
```diff
@@ -522,28 +525,52 @@ struct ad77681_dev {
...
```
Comment (cristiantpop) [inline ad7768-1fmcz/ad77681.h:561]: Same here

File: ad7768-1fmcz/ad77681.c [hunk 3]
```diff
@@ -237,6 +237,35 @@ uint8_t ad77681_get_rx_buf_len(struct ad77681_dev *dev)
 	return rx_buf_len;
 }
 
+/**
+ * Helper function to get the number of SPI 16bit frames for INTERRUPT ADC DATA READ
+ * @param dev - The device structure.
+ * @return frame_16bit - the number of 16 bit SPI frames
+ */
+uint8_t ad77681_get_frame_16bit(struct ad77681_dev *dev)
```
Comment (cristiantpop) [inline ad7768-1fmcz/ad77681.c:245]: Add a description in the commit message, why this is needed and where

File: ad7768-1fmcz/ad77681.c [hunk 4]
```diff
@@ -351,14 +354,17 @@ int32_t ad77681_set_power_mode(struct ad77681_dev *dev,
...
-			       AD77681_POWER_CLK_MCLK_DIV(clk_div));
+	int32_t ret;
+
+	ret = ad77681_spi_write_mask(dev,
+				     AD77681_REG_POWER_CLOCK,
+				     AD77681_POWER_CLK_MCLK_DIV_MSK,
+				     AD77681_POWER_CLK_MCLK_DIV(clk_div));
 
-	dev->mclk_div = clk_div;
+	if (ret==SUCCESS)
```
Comment (cristiantpop) [inline ad7768-1fmcz/ad77681.c:645]: Add some spaces, like this: `if (ret == SUCCESS)`

File: ad7768-1fmcz/ad77681.c [hunk 5]
```diff
@@ -316,6 +316,287 @@ int32_t ad77681_spi_read_adc_data(struct ad77681_dev *dev,
...
+	/* 24bit format*/
+	if (dev->conv_len == AD77681_CONV_24BIT)
+		/* Cuts first 8 bits from buf[1] and paste it after 16bits of buf[0]*/
+		measured_data->raw_data[measured_data->count] = (buf[0] << 8 | buf[1] >> 8);
+	/* 16bit format */
+	else
+		measured_data->raw_data[measured_data->count] = (buf[0] << 8);
+
+	/* If status bit or CRC is enabled, handle it */
+	if(dev->status_bit || (dev->crc_sel != AD77681_NO_CRC))
```
Comment (cristiantpop) [inline ad7768-1fmcz/ad77681.c:359]: Add a space between if and its condition: `if (dev->status_bit || (dev->crc_sel != AD77681_NO_CRC))`
Check every if statement

File: ad7768-1fmcz/ad77681.c [hunk 6]
```diff
@@ -316,6 +316,287 @@ int32_t ad77681_spi_read_adc_data(struct ad77681_dev *dev,
...
+						   measured_data->raw_data[measured_data->count]);
+		else
+			converted_data = (int32_t)((0x00 << 24) |
+						   measured_data->raw_data[measured_data->count]);
+
+		/* ((2*Vref)*code)/2^24	*/
+		voltage[measured_data->count] = (double)(((2.0 * (((double)(
+							dev->vref)) / 1000.0)) / AD7768_FULL_SCALE)*converted_data);
+	}
+	return 0;
```
Comment (cristiantpop) [inline ad7768-1fmcz/ad77681.c:471]: Add extra line before return.

File: ad7768-1fmcz/ad77681.c [hunk 7]
```diff
@@ -287,6 +316,287 @@ int32_t ad77681_spi_read_adc_data(struct ad77681_dev *dev,
...
+ * 8bits of status bit, if enabled	dev->sataus_bit    followed by
+ * 8bits of CRC, if enabled			dev->crc_sel
+ *
+ * @param dev - The device structure.
+ * @param measured_data - structure carying ADC measurement info
+ * @return 0 in case of success, negative error code otherwise.
+ */
+/*
+*	    /-------------\       /-------------\       /-------------\
+*------  24-bit data  -------  8-bit staus  -------   8-bit CRC   -------
```
Comment (amiclaus) [inline ad7768-1fmcz/ad77681.c:332]: I would not use this type of comments (sequence drawing with characters). The doxygen build will most probably mess it up. Maybe you can reference this sequence from the datasheet.

File: ad7768-1fmcz/ad77681.c [hunk 8]
```diff
@@ -486,21 +1105,555 @@ int32_t ad77681_soft_reset(struct ad77681_dev *dev)
...
 
 	dev = (struct ad77681_dev *)malloc(sizeof(*dev));
 	if (!dev) {
-		return -1;
+		return FAILURE;
+	}
+
+	stat = (struct ad77681_status_registers *)malloc(sizeof(*stat));
+	if (!stat) {
+		return FAILURE;
```
Comment (amiclaus) [inline ad7768-1fmcz/ad77681.c:1656]: `free(dev);` before returning FAILURE

File: ad7768-1fmcz/ad77681.c [hunk 9]
```diff
@@ -322,12 +635,306 @@ int32_t ad77681_set_power_mode(struct ad77681_dev *dev,
...
+ * @return 0 in case of success, negative error code otherwise.
+ */
+
+int32_t ad77681_power_down(struct ad77681_dev *dev,
+			   enum ad77681_sleep_wake sleep_wake)
+{
+	int32_t ret;
+
+	if (sleep_wake == AD77681_SLEEP) {
+		ret = ad77681_spi_reg_write(dev, AD77681_REG_POWER_CLOCK,
```
Comment (amiclaus) [inline ad7768-1fmcz/ad77681.c:927]: the value of `ret` is not handled in this function.
Replace `return 0` with `return ret` at the end of the function

File: ad7768-1fmcz/ad77681.c [hunk 10]
```diff
@@ -287,6 +316,287 @@ int32_t ad77681_spi_read_adc_data(struct ad77681_dev *dev,
...
+		if (measured_data->raw_data[measured_data->count] & 0x800000)
+			converted_data = (int32_t)((0xFF << 24) |
+						   measured_data->raw_data[measured_data->count]);
+		else
+			converted_data = (int32_t)((0x00 << 24) |
+						   measured_data->raw_data[measured_data->count]);
+
+		/* ((2*Vref)*code)/2^24	*/
+		voltage[measured_data->count] = (double)(((2.0 * (((double)(
+							dev->vref)) / 1000.0)) / AD7768_FULL_SCALE)*converted_data);
```
Comment (amiclaus) [inline ad7768-1fmcz/ad77681.c:469]: nitpick: add spaces:
`AD7768_FULL_SCALE) * converted_data`

File: ad7768-1fmcz/ad77681.c [hunk 11]
```diff
@@ -287,6 +316,287 @@ int32_t ad77681_spi_read_adc_data(struct ad77681_dev *dev,
...
+
+		for (i = 0; i < checksum_length; i++) {
+			if (i % 2)
+				checksum_buf[i] = data_buffer[i / 2] & 0xFF;
+			else
+				checksum_buf[i] = data_buffer[i / 2] >> 8;
+		}
+
+		if (dev->crc_sel == AD77681_CRC)
+			checksum = ad77681_compute_crc8(checksum_buf, checksum_length,INITIAL_CRC_CRC8);
```
Comment (amiclaus) [inline ad7768-1fmcz/ad77681.c:425]: nitpick: add space
`checksum_length, INITIAL_CRC_CRC8`

**Comment by amiclaus** [comment]
Also, have a look at the build issues detected by TravisCI and Codacy.
And please try to fix them.

File: ad7768-1fmcz/ad77681.c (hunk 3 above)
Comment (jgarlitos) [reply ad7768-1fmcz/ad77681.c:245]: Done adding description on the commit

File: ad7768-1fmcz/ad77681.c (hunk 4 above)
Comment (jgarlitos) [reply ad7768-1fmcz/ad77681.c]: Done fixing

File: ad7768-1fmcz/ad77681.c (hunk 5 above)
Comment (jgarlitos) [reply ad7768-1fmcz/ad77681.c]: Done fixing

File: ad7768-1fmcz/ad77681.c (hunk 6 above)
Comment (jgarlitos) [reply ad7768-1fmcz/ad77681.c]: Done fixing

File: ad7768-1fmcz/ad77681.c (hunk 7 above)
Comment (jgarlitos) [reply ad7768-1fmcz/ad77681.c]: done removing thin comment

File: ad7768-1fmcz/ad77681.c (hunk 8 above)
Comment (jgarlitos) [reply ad7768-1fmcz/ad77681.c]: Memory leak build error was fixed when returning -1 instead of failure.

File: ad7768-1fmcz/ad77681.c (hunk 9 above)
Comment (jgarlitos) [reply ad7768-1fmcz/ad77681.c:911]: Done fixing

File: ad7768-1fmcz/ad77681.c (hunk 11 above)
Comment (jgarlitos) [reply ad7768-1fmcz/ad77681.c]: fixed

**Comment by jgarlitos** [comment]
Hi All,
Done fixing build errors and implementing review comments.

**Comment by jgarlitos** [comment]
Hi All,
I had multiple force commits while fixing build issues detected by TravisCI and Codacy in this pull request.
Would it be better to close this pull request and initiate another one to have a fresh commits history on the pull request?

File: ad7768-1fmcz/ad77681.c [hunk 12]
```diff
@@ -316,6 +316,271 @@ int32_t ad77681_spi_read_adc_data(struct ad77681_dev *dev,
...
+	if (dev->status_bit || (dev->crc_sel != AD77681_NO_CRC))
+		ret |= ad77681_CRC_status_handling(dev, buf);
+
+	return ret;
+}
+
+/**
+ * CRC and status bit handling after each readout form the ADC
+ * @param dev - The device structure.
+ * @param *data_buffer - 16-bit buffer readed from the ADC containing the CRC, data and the stattus bit
```
Comment (amiclaus) [inline ad7768-1fmcz/ad77681.c:362]: nitpick: max char length should be 80

File: ad7768-1fmcz/ad77681.c [hunk 13]
```diff
@@ -316,6 +316,271 @@ int32_t ad77681_spi_read_adc_data(struct ad77681_dev *dev,
...
+	}
+
+	/* Sample rate to Hz */
+	dev->sample_rate = (dev->mclk / (osr*mclk_div)) * 1000;
+
+	return SUCCESS;
+}
+
+/**
+ * Get SINC3 filter oversampling ratio register value based on user's inserted output data rate ODR
```
Comment (amiclaus) [inline ad7768-1fmcz/ad77681.c:539]: nitpick: max char length should be 80

File: ad7768-1fmcz/ad77681.c [hunk 14]
```diff
@@ -367,6 +632,297 @@ int32_t ad77681_set_mclk_div(struct ad77681_dev *dev,
 	return ret;
 }
 
+/**
+ * Set the VCM output.
+ * @param dev - The device structure.
+ * @param VCM_out - The VCM output voltage.
+ * 					Accepted values: AD77681_VCM_HALF_VCC
+ *									 AD77681_VCM_2_5V
+	  								 AD77681_VCM_2_05V
```
Comment (amiclaus) [inline ad7768-1fmcz/ad77681.c:641]: nitpick: This header comment looks messy.

File: ad7768-1fmcz/ad77681.c [hunk 15]
```diff
@@ -367,6 +632,297 @@ int32_t ad77681_set_mclk_div(struct ad77681_dev *dev,
...
+		dev->VCM_out = VCM_out;
+
+	return ret;
+}
+
+/**
+ * Set the AIN- precharge buffer.
+ * @param dev - The device structure.
+ * @param AINn - The negative analog input precharge buffer selector
+ * 					Accepted values: AD77681_AINn_ENABLED
```
Comment (amiclaus) [inline ad7768-1fmcz/ad77681.c:669]: nitpick: This header comment looks messy. Check the entire file for similar issues.

File: ad7768-1fmcz/ad77681.c (hunk 12 above)
Comment (jgarlitos) [reply ad7768-1fmcz/ad77681.c]: Done.

File: ad7768-1fmcz/ad77681.c (hunk 14 above)
Comment (jgarlitos) [reply ad7768-1fmcz/ad77681.c]: Done.

File: ad7768-1fmcz/ad77681.c (hunk 13 above)
Comment (jgarlitos) [reply ad7768-1fmcz/ad77681.c]: Done.

File: ad7768-1fmcz/ad77681.c (hunk 15 above)
Comment (jgarlitos) [reply ad7768-1fmcz/ad77681.c]: Done.

**Comment by jgarlitos** [comment]
Done fixing issues.

**Comment by jgarlitos** [comment]
Hi All,
I don't know why the "Travis CI build" is still in progress. I would like to ask assistance on this matter.
Thanks.

**Comment by amiclaus** [comment]
Hello,

I've re-triggered the Travis build and it is finished now.

There are still some issues with the code style. I suggest performing a rebase with the latest version of the master branch.

Also try to run the code style check locally to see if there are any issues: https://github.com/analogdevicesinc/no-OS/wiki/Code-Style-guidelines

**Comment by jgarlitos** [comment]
Done re-basing and error was fixed.

**Comment by jgarlitos** [comment]
Hi All,
Is there something I need to do in pull request?
Thanks

**Comment by jgarlitos** [comment]
Done fixing merge conflicts.

**Comment by jgarlitos** [comment]
Hi all,
I'm not sure why the pull request failed in codacy due to memory leak.

File: drivers/adc/ad7768-1/ad77681.c [hunk 16]
```diff
@@ -494,16 +1620,25 @@ int32_t ad77681_soft_reset(struct ad77681_dev *dev)
...
+	uint8_t scratchpad_check = 0xAD;
 
 	dev = (struct ad77681_dev *)malloc(sizeof(*dev));
 	if (!dev) {
 		return -1;
 	}
 
+	stat = (struct ad77681_status_registers *)malloc(sizeof(*stat));
+	if (!stat) {
+		free(stat);
```
Comment (amiclaus) [inline drivers/adc/ad7768-1/ad77681.c:1638]: `free(dev)` instead of `free(stat)`.

This should fix the Codacy issue.

Comment (amiclaus) [reply drivers/adc/ad7768-1/ad77681.c:1638]: When you reach this point in the code, only `dev` has memory successfully allocated. So before returning -1 it needs to be freed.

File: drivers/adc/ad7768-1/ad77681.c [hunk 17]
```diff
@@ -288,6 +317,273 @@ int32_t ad77681_spi_read_adc_data(struct ad77681_dev *dev,
...
+#endif /* CRC_DEBUG */
+	}
+
+	return ret;
+}
+
+/**
+ * Conversion from measured data to voltage
+ * @param dev - The device structure.
+ * @param measured_data - structure carying ADC measurement info
```
Comment (amiclaus) [inline drivers/adc/ad7768-1/ad77681.c:437]: `ad77681_data_to_voltage` function does not have `measured_data` as parameter. This was also detected by the doxygen build (TravisCI). There are similar issues throughout the code (check the Travis build).

File: drivers/adc/ad7768-1/ad77681.c (hunk 16 above)
Comment (jgarlitos) [reply drivers/adc/ad7768-1/ad77681.c]: Thanks. Done fixing.

File: drivers/adc/ad7768-1/ad77681.c (hunk 17 above)
Comment (jgarlitos) [reply drivers/adc/ad7768-1/ad77681.c]: Thanks. Done fixing but, no build results from Travis.

**Comment by jgarlitos** [comment]
Done fixing build error.

**Pull request merged**